proposals/handler.go

```diff
@@ -0,0 +1,380 @@
+	errConflictingExceptions = errors.New("conflicting exceptions")
+	errExceptionsOverflow    = errors.New("too many exceptions")
+	errDuplicateTX           = errors.New("duplicate TxID in proposal")
+	errDuplicateATX          = errors.New("duplicate ATXID in active set")
+)
+
+// Handler processes Proposal from gossip and, if deems it valid, propagates it to peers.
+type Handler struct {
+	logger log.Log
+	cfg    Config
+
+	fetcher   system.Fetcher
+	mesh      mesh
+	validator eligibilityValidator
+}
+
+// Config defines configuration for the handler.
+type Config struct {
+	LayerSize            uint32
+	LayersPerEpoch       uint32
+	GoldenATXID          types.ATXID
+	MaxExceptions        int
+	EligibilityValidator eligibilityValidator
+}
+
+// defaultConfig for BlockHandler.
+func defaultConfig() Config {
+	return Config{
+		MaxExceptions:        1000,
+		EligibilityValidator: nil,
+	}
+}
+
+// Opt for configuring BlockHandler.
+type Opt func(h *Handler)
+
+// WithGoldenATXID defines the golden ATXID.
+func WithGoldenATXID(atx types.ATXID) Opt {
+	return func(h *Handler) {
+		h.cfg.GoldenATXID = atx
+	}
+}
+
+// WithLayerSize defines the average number of proposal per layer.
+func WithLayerSize(size uint32) Opt {
+	return func(h *Handler) {
+		h.cfg.LayerSize = size
+	}
+}
+
+// WithLayerPerEpoch defines the number of layers per epoch.
+func WithLayerPerEpoch(layers uint32) Opt {
+	return func(h *Handler) {
+		h.cfg.LayersPerEpoch = layers
+	}
+}
+
+// WithValidator defines eligibility validator.
+func WithValidator(v eligibilityValidator) Opt {
+	return func(h *Handler) {
+		h.validator = v
+	}
+}
+
+// WithMaxExceptions defines max allowed exceptions in a ballot.
+func WithMaxExceptions(max int) Opt {
+	return func(h *Handler) {
+		h.cfg.MaxExceptions = max
+	}
+}
+
+// WithLogger defines logger for Handler.
+func WithLogger(logger log.Log) Opt {
+	return func(h *Handler) {
+		h.logger = logger
+	}
+}
+
+// NewHandler creates new Handler.
+func NewHandler(f system.Fetcher, bc system.BeaconCollector, db atxDB, m mesh, opts ...Opt) *Handler {
+	b := &Handler{
+		logger:  log.NewNop(),
+		cfg:     defaultConfig(),
+		fetcher: f,
+		mesh:    m,
+	}
+	for _, opt := range opts {
+		opt(b)
+	}
+	if b.validator == nil {
+		b.validator = newEligibilityValidator(b.cfg.LayerSize, b.cfg.LayersPerEpoch, db, bc, m, b.logger)
+	}
+	return b
+}
+
+// HandleProposal is the gossip receiver for Proposal.
+// TODO: register this gossip handler.
+func (h *Handler) HandleProposal(ctx context.Context, _ peer.ID, msg []byte) pubsub.ValidationResult {
+	if err := h.processProposal(ctx, msg); err != nil {
+		h.logger.WithContext(ctx).With().Error("failed to process proposal gossip", log.Err(err))
+		return pubsub.ValidationIgnore
+	}
+	return pubsub.ValidationAccept
+}
+
+// HandleBallot is the gossip receiver for Ballot.
+// TODO: register this gossip handler.
+func (h *Handler) HandleBallot(ctx context.Context, _ peer.ID, msg []byte) pubsub.ValidationResult {
```

Comment:
why do we gossip both but proposal also has a ballot included?

Reply:
sorry. this was a mistake. we don't gossip Ballot separately. we will only fetch Ballot directly from peers.